provider/consulcatalog/consul_catalog.go

```diff
@@ -156,14 +155,8 @@ func (p *Provider) watch(configurationChan chan<- types.ConfigMessage, stop chan
 	defer close(stopCh)
 	defer close(watchCh)
 
-	for {
-		select {
-		case <-stop:
-			return nil
-		case index, ok := <-watchCh:
-			if !ok {
-				return errors.New("consul service list nil")
-			}
+	go func() {
```

Comment:
You can't just launch new goroutine without handling the stop. Because we need to be able to stop all providers goroutines when we want. This is the purpose of the `safe.Pool` we receive in the `Provide` method, and this is the purpose of the stop chan here.

Reply:
Not really familiar with Traefik's design, but the stop channel will cause the watch method to return, causing `defer close(watchCh)` [here](https://github.com/containous/traefik/pull/3560/files/257d278191f547db00c5cd1fccd008324e5b8a60#diff-cd8329c0a91a00757f75bd879fa2cf7cL157)  to execute, which will in turn end the iteration in the goroutine - so stop is handled implicitly. Or am I misunderstanding?

Reply:
You're right, it will be closed, but I think we should use a `safe.Go` (to recover eventual panic)